File: web/src/types/progression.ts
export interface PendingLevelReward {
    level: number;
    reward_resource_id: string | null;
    reward_amount: number | null;
    reward_booster_id: string | null;
    reward_booster_name: string | null;
    /** Niveau de prestige (0 = route normale) et son bonus. */
    prestige: number;
    bonus_resource_id: string | null;
    bonus_amount: number | null;
}

export interface LevelStatus {
    current_level: number;
    prestige: number;
    total_power: number;
    claimed_level: number;
    /** Seuil du niveau atteint : la barre se remplit ENTRE les deux seuils,
     * sinon elle mesure le chemin depuis zéro et reste presque pleine. */
    current_level_power_required: number;
    next_level_power_required: number | null;
    pending_rewards: PendingLevelReward[];
    has_unclaimed: boolean;
}

export interface LevelTierOverview {
    level: number;
    power_required: number;
    reward_resource_id: string | null;
    reward_resource_name: string | null;
    reward_amount: number | null;
    reward_booster_id: string | null;
    reward_booster_name: string | null;
    prestige: number;
    bonus_resource_id: string | null;
    bonus_resource_name: string | null;
    bonus_amount: number | null;
    reached: boolean;
    claimed: boolean;
}

export type AchievementCategory = "collection" | "social" | "economy" | "progression" | "meta";

export interface Achievement {
    id: string;
    name: string;
    description: string;
    category: AchievementCategory;
    threshold: number;
    progress: number;
    reward_resource_id: string | null;
    reward_amount: number | null;
    reward_booster_id: string | null;
    unlocked_at: string | null;
    claimed_at: string | null;
}

export type QuestPeriod = "daily" | "weekly";

export interface Quest {
    id: number;
    quest_def_id: string;
    name: string;
    description: string;
    period: QuestPeriod;
    threshold: number;
    progress: number;
    reward_resource_id: string | null;
    reward_amount: number | null;
    reward_booster_id: string | null;
    completed: boolean;
    claimed_at: string | null;
}

File: backend/app/schemas/progression.py
"""
Schemas — niveaux, achievements, quêtes.
"""

from datetime import datetime
from pydantic import BaseModel


class PendingLevelReward(BaseModel):
    level: int
    reward_resource_id: str | None = None
    reward_amount: int | None = None
    reward_booster_id: str | None = None
    reward_booster_name: str | None = None
    # Niveaux de prestige (au-delà de la route) : numéro et bonus.
    prestige: int = 0
    bonus_resource_id: str | None = None
    bonus_amount: int | None = None


class LevelStatus(BaseModel):
    current_level: int
    prestige: int = 0  # 0 tant que la route n'est pas terminée
    total_power: int
    claimed_level: int
    next_level_power_required: int | None = None
    pending_rewards: list[PendingLevelReward] = []
    has_unclaimed: bool = False


class LevelTierOut(BaseModel):
    """Un palier de la table complète — pour la "route" affichant tous les
    niveaux avec leurs récompenses (cf. Progression.tsx)."""
    level: int
    power_required: int
    reward_resource_id: str | None = None
    reward_resource_name: str | None = None
    reward_amount: int | None = None
    reward_booster_id: str | None = None
    reward_booster_name: str | None = None
    prestige: int = 0
    bonus_resource_id: str | None = None
    bonus_resource_name: str | None = None
    bonus_amount: int | None = None
    reached: bool = False
    claimed: bool = False


class AchievementOut(BaseModel):
    id: str
    name: str
    description: str
    category: str
    threshold: int
    progress: int
    reward_resource_id: str | None = None
    reward_amount: int | None = None
    reward_booster_id: str | None = None
    unlocked_at: datetime | None = None
    claimed_at: datetime | None = None


class QuestOut(BaseModel):
    id: int
    quest_def_id: str
    name: str
    description: str
    period: str
    threshold: int
    progress: int
    reward_resource_id: str | None = None
    reward_amount: int | None = None
    reward_booster_id: str | None = None
    completed: bool
    claimed_at: datetime | None = None
